File: src/main/java/pt/joaomneto/titancompanion/adventure/SpellAdventure.java
package pt.joaomneto.titancompanion.adventure;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import pt.joaomneto.titancompanion.adventure.impl.util.Spell;

public abstract class SpellAdventure extends Adventure {


    List<Spell> spells = new ArrayList<Spell>();;


    protected String arrayToStringSpells(List<Spell> spells) {
        String _string = "";

        if (!spells.isEmpty()) {
            for (Spell spell : spells) {
                _string += spell+ "#";
            }
            _string = _string.substring(0, _string.length() - 1);
        }
        return _string;
    }

    protected List<Spell> stringToArraySpells(String spells) {
        List<Spell> elements = new ArrayList<Spell>();

        if (spells != null) {
            List<String> list = Arrays.asList(spells.split("#"));
            for (String string : list) {
                if (!string.isEmpty()) {
                        elements.add(getSpellForString(string));
                }
            }
        }

        return elements;
    }

    private Spell getSpellForString(String string) {
        try {
            String simpleName = this.getClass().getSimpleName();
            simpleName = simpleName.substring(0, simpleName.length() - "Adventure".length());
            Class<? extends Spell> clazz = (Class<? extends Spell>) Class.forName("pt.joaomneto.titancompanion.adventure.impl.fragments."+simpleName.toLowerCase()+ "." + simpleName + "Spell");
            Method method = clazz.getMethod("valueOf", String.class);
            Spell spell = (Spell) method.invoke(null, string);
            return spell;
        } catch (Exception e) {
            throw new IllegalArgumentException("Unable to instantiate Spell enum for class"+this.getClass(), e);
        }
    }

    public List<Spell> getSpells() {
        return spells;
    }

    public void setSpells(List<Spell> spells) {
        this.spells = spells;
    }

    public abstract List<Spell> getSpellList();

    public abstract boolean isSpellSingleUse();
}

File: src/main/java/pt/joaomneto/titancompanion/adventure/SpellAdventure.kt
//package pt.joaomneto.titancompanion.adventure
//
//import pt.joaomneto.titancompanion.adventure.impl.util.Spell
//import pt.joaomneto.titancompanion.util.AdventureFragmentRunner
//import java.util.Arrays
//
//abstract class SpellAdventure<S : Spell>(override val fragmentConfiguration: Array<AdventureFragmentRunner>) : Adventure(
//    fragmentConfiguration
//) {
//
//    var chosenSpells: List<S> = emptyList()
//    abstract val spellList: List<S>
//    abstract val isSpellSingleUse: Boolean
//
//    protected fun arrayToStringSpells(spells: List<S>): String {
//        var _string = ""
//
//        if (!spells.isEmpty()) {
//            for (spell in spells) {
//                _string += spell.toString() + "#"
//            }
//            _string = _string.substring(0, _string.length - 1)
//        }
//        return _string
//    }
//
//    protected fun stringToArraySpells(spells: String?, type: Class<S>): MutableList<S> {
//        val elements = ArrayList<S>()
//
//        if (spells != null) {
//            val list = Arrays.asList(*spells.split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
//            for (string in list) {
//                if (!string.isEmpty()) {
//                    elements.add(getSpellForString(string, type))
//                }
//            }
//        }
//
//        return elements
//    }
//
//    private fun <S> getSpellForString(string: String, type: Class<S>): S {
//        try {
//            val method = type.getMethod("valueOf", String::class.java)
//            return method.invoke(null, string) as S
//        } catch (e: Exception) {
//            throw IllegalArgumentException("Unable to instantiate Spell enum for class" + this.javaClass, e)
//        }
//    }
//}
